include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -673,241 +500,442 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, NdotL2, a2);
+			ndf_val = irr_glsl_beckmann(a2, NdotH2);
 #else
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams, currInteraction, n, ny, a2, ay2);
-			pdf = irr_glsl_beckmann_pdf(s, currInteraction, a, a2, ay, ay2);
+			float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, a2, ay2);
+			G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL * s.TdotL, s.BdotL * s.BdotL, NdotL2, a2, ay2);
+			ndf_val = irr_glsl_beckmann(ax, ay, a2, ay2, TdotH2, BdotH2, NdotH2);
 #endif
-
-		} else
+		} CASE_END
 #endif
-
-#ifndef ONLY_ONE_NDF
+		CASE_OTHERWISE
 		{} //else "empty braces"
-#endif
-	}
+		END_CASES
 
-	uvec3 regs = instr_decodeRegisters(instr);
-#ifdef OP_DIFFUSE
-	if (op==OP_DIFFUSE) {
-		instr_execute_cos_eval_DIFFUSE(instr, regs, params, bsdf_data);
-	} else
-#endif
+		bxdf_eval_scalar_part = 0.5 * G2_over_G1 * G1_over_2NdotV * ndf_val; //cos*D*G2/(4*cos*cos)
+		pdf = irr_glsl_smith_VNDF_pdf_wo_clamps(ndf_val, G1_over_2NdotV);
+
+		float VdotH = irr_glsl_conditionalAbsOrMax(is_bsdf, microfacet.isotropic.VdotH, 0.0);
+		vec3 fr;
 #ifdef OP_CONDUCTOR
-	if (op==OP_CONDUCTOR) {
-		instr_execute_cos_eval_CONDUCTOR(instr, regs, bxdf_eval_scalar_part, params, bsdf_data);
-	} else
+		if (op == OP_CONDUCTOR)
+			fr = irr_glsl_fresnel_conductor(ior[0], ior[1], VdotH);
+		else
 #endif
-#ifdef OP_PLASTIC
-	if (op==OP_PLASTIC) {
-		pdf = instr_execute_cos_eval_pdf_PLASTIC(instr, regs, bxdf_eval_scalar_part, params, bsdf_data, s, pdf);
-	} else
+			fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH);
+
+#ifndef NO_BSDF
+		if (is_bsdf && is_valid)
+		{
+			float eta = ior_scalar;
+
+			float LdotH = microfacet.isotropic.LdotH;
+			float VdotHLdotH = microfacet.isotropic.VdotH * LdotH;
+			LdotH = abs(LdotH);
+#ifdef NDF_GGX
+			if (ndf == NDF_GGX)
+				bxdf_eval_scalar_part = irr_glsl_ggx_microfacet_to_light_measure_transform(bxdf_eval_scalar_part, NdotL, refraction, VdotH, LdotH, VdotHLdotH, eta);
+			else
 #endif
-#ifdef OP_COATING
-	if (op==OP_COATING) {
-		instr_execute_cos_eval_COATING(instr, regs, params, bsdf_data);
-	} else
+				bxdf_eval_scalar_part = irr_glsl_microfacet_to_light_measure_transform(bxdf_eval_scalar_part, NdotV, refraction, VdotH, LdotH, VdotHLdotH, eta);
+		}
 #endif
-#ifdef OP_DIFFTRANS
-	if (op==OP_DIFFTRANS) {
-		instr_execute_cos_eval_DIFFTRANS(instr, regs, params, bsdf_data);
-	} else
+			
+		bxdf_eval = is_valid ? vec3(bxdf_eval_scalar_part) : vec3(0.0);
+
+#ifdef OP_CONDUCTOR
+		if (op == OP_CONDUCTOR)
+			bxdf_eval *= fr;
+		else
 #endif
-#ifdef OP_DIELECTRIC
-	if (op==OP_DIELECTRIC) {
-		instr_execute_cos_eval_DIELECTRIC(instr, regs, params, bsdf_data);
+		{
+			float reflectance = dot(CIE_XYZ_Luma_Y_coeffs, fr);
+			reflectance = refraction ? (1.0 - reflectance) : reflectance;
+			pdf *= reflectance;
+			bxdf_eval *= reflectance;
```

Comment:
listen to me...
and only compute the remainder
https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r526500143
mul by pdf at the end!

Reply:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r526501255